### PR TITLE
fix(reproducibility): add opt-in strict determinism across trainers

### PR DESCRIPTION
**Pull request by SoheylM**
# Description
Adds an opt-in strict reproducibility path for all training entrypoints while preserving current default behavior.

- Introduces `engiopt/reproducibility.py` with shared helpers:
  - `seed_training(seed)`
  - `enable_strict_determinism(warn_only=True)`
  - `make_dataloader_generator(seed)`
- Adds `strict_determinism: bool = False` to all targeted training `Args` dataclasses.
- Replaces ad-hoc seeding with `seed_training(args.seed)` in each training script.
- Enables strict deterministic controls only when `--strict-determinism` is passed.
- Hardens DataLoader reproducibility by supplying seeded generators for `shuffle=True` loaders.
- Updates README with optional reproducibility usage (`--strict-determinism`).

Fixes SOH-14 (Linear)

## Type of change

- [ ] Documentation only change (no code changed)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New algorithm (non-breaking change which adds a new model/algorithm)
- [x] Improvement to existing algorithm (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### Screenshots
N/A

# Checklist:

### Code Quality
- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files`
- [x] I have run `ruff check .` and `ruff format`
- [x] I have run `mypy .`
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

### CleanRL Philosophy (for new/modified algorithms)
- [x] The implementation follows the [CleanRL](https://github.com/vwxyzjn/cleanrl) single-file philosophy: all training logic is contained in one file
- [x] The code is reproducible: random seeds are set, PyTorch determinism is enabled
- [x] Hyperparameters are configurable via command-line arguments using `tyro`
- [x] WandB logging is integrated with `--track` flag support
- [x] The model can be saved and restored via WandB artifacts (`--save-model` flag)

### Algorithm Completeness (for new algorithms)
- [ ] Both training script (`algorithm.py`) and evaluation script (`evaluate_algorithm.py`) are provided
- [ ] The algorithm works with EngiBench's `Problem` interface
- [ ] The algorithm is added to the README table with correct metadata

### Documentation
- [x] I have made corresponding changes to the documentation
- [ ] New algorithms include docstrings explaining the approach and any paper references

## Validation
- Determinism smoke check (`cgan_cnn_2d`) run twice on same machine with strict mode and fixed seed; resulting checkpoint tensor hashes matched for both generator and discriminator.
- Non-strict short run completed successfully (default behavior path unchanged).
- Strict mode configured with `warn_only=True` for nondeterministic ops to warn and continue.

### CODE REVIEW

File: engiopt/gan_bezier/gan_bezier.py [hunk 1]
```diff
@@ -427,7 +432,19 @@ def prepare_data(problem, batch_size, device):
...
+            batch_size=batch_size,
+            shuffle=True,
+        )
+    else:
+        dataloader = th.utils.data.DataLoader(
+            training_ds,
+            batch_size=batch_size,
+            shuffle=True,
+            generator=make_dataloader_generator(seed),
+        )
```
Comment (g-braeunlich) [inline engiopt/gan_bezier/gan_bezier.py]:
```suggestion
    dataloader = th.utils.data.DataLoader(
        training_ds,
        batch_size=batch_size,
        shuffle=True,
        generator=make_dataloader_generator(seed) if seed is not None else None,
    )
```

File: engiopt/reproducibility.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    return rng
+
+
+def enable_strict_determinism(*, warn_only: bool = True) -> None:
+    """Enable stricter deterministic settings without changing model logic."""
+    os.environ.setdefault("CUBLAS_WORKSPACE_CONFIG", ":4096:8")
+    torch.backends.cudnn.deterministic = True
+    torch.backends.cudnn.benchmark = False
+    if torch.cuda.is_available():
+        torch.backends.cuda.matmul.allow_tf32 = False
```
Comment (g-braeunlich) [inline engiopt/reproducibility.py]: Not exactly the same logic, but maybe safer?

```suggestion
    torch.backends.cuda.matmul.allow_tf32 = not torch.cuda.is_available()
```

**Review by g-braeunlich — COMMENTED** [review]
Just 2 small suggestions

File: engiopt/gan_bezier/gan_bezier.py (hunk 1 above)
Comment (SoheylM) [reply engiopt/gan_bezier/gan_bezier.py]: Applied, thanks. I pushed commit 21e760f, which uses a single DataLoader call with `generator=make_dataloader_generator(seed) if seed is not None else None` (same behavior, cleaner wiring).

File: engiopt/reproducibility.py (hunk 2 above)
Comment (SoheylM) [reply engiopt/reproducibility.py]: I kept the current logic intentionally for strict mode semantics. In strict mode we should avoid enabling TF32; `not torch.cuda.is_available()` would set this flag to `True` on CPU/MPS runs. Current behavior only disables CUDA matmul TF32 when CUDA exists, while CPU/MPS remain unaffected. If we simplify, I’d still keep the assigned value `False` and only guard backend availability.

Comment (SoheylM) [reply engiopt/reproducibility.py]: Follow-up: I agree the cleaner framing is backend-capability based. For strict mode, we should keep TF32 disabled (`False`) whenever the CUDA matmul backend is present, rather than deriving the value from `torch.cuda.is_available()` (which can be `False` on CPU/MPS runs and would imply `True`). Happy to switch to that form for clarity if you prefer.

Comment (SoheylM) [reply engiopt/reproducibility.py]: Implemented in commit f11c598: I changed the guard to backend capability (`hasattr(torch.backends, "cuda") and hasattr(torch.backends.cuda, "matmul")`) while keeping strict-mode TF32 set to `False`. This keeps CPU/MPS unaffected and avoids deriving strict behavior from runtime CUDA availability. Could you confirm yes/no if this addresses your point?

**Comment by SoheylM** [comment]
@g-braeunlich quick ping when you have a moment: could you confirm yes/no on the latest TF32 guard update in the open thread? If yes, we’ll resolve and merge. Thanks!

**Review by g-braeunlich — APPROVED** [review]
